deployment/ccip/changeset: add transfer/accept ownership changeset

makramkd opened the pull request:
Ticket: https://smartcontract-it.atlassian.net/browse/CCIP-4396

This PR adds:
* TransferOwnership changeset: this changeset would get executed after an initial deploy of all the contracts, that transfers ownership of them to the appropriate timelock on each chain. 
* AcceptOwnership changset: this changeset would get executed after TransferOwnership, and it generates the accept ownership proposal.

This PR also adds a test for the above changesets, and deletes some TODO'd code that does the same stuff as the above and updates the tests to use the changesets instead.

makramkd committed:
wip

makramkd committed:
Merge branch 'develop' into mk/accept-ownership-proposal

makramkd committed:
add test

makramkd committed:
make the proposal multichain

makramkd committed:
add transfer ownership changeset

use these changesets in other tests and axe duplicate code

makramkd committed:
fix add_chain_test.go

makramkd committed:
extract common code to func

AnieeG commented on deployment/ccip/changeset/accept_ownership.go:
```diff
@@ -0,0 +1,149 @@
+}
+
+type AcceptOwnershipConfig struct {
+	State             CCIPOnChainState
```
is this required here? It's difficult to pass the state from chainlink-deployments cs application, you have the environment. can you form the state within the changeset func?

makramkd replied:
I've since removed this due to moving to common, but I noticed at least 3 other changesets that take in the CCIP onchain state directly

makramkd replied:
Created a ticket to address here: https://smartcontract-it.atlassian.net/browse/CCIP-4394

AnieeG commented on deployment/ccip/changeset/accept_ownership.go:
```diff
@@ -0,0 +1,149 @@
+	cfg AcceptOwnershipConfig,
+) (deployment.ChangesetOutput, error) {
+	// basic validation
+	if len(cfg.ChainSelectors) == 0 || cfg.HomeChainSelector == 0 {
+		return deployment.ChangesetOutput{}, fmt.Errorf("no chain selectors provided")
+	}
+
+	if len(cfg.State.Chains) == 0 {
+		return deployment.ChangesetOutput{}, fmt.Errorf("no chains in state")
+	}
```
better to create a validate func following the changeset [pattern](https://github.com/smartcontractkit/chainlink/blob/develop/deployment/ccip/changeset/deploy_chain.go#L19-L21). It will be easier in future if we need to add more validation/more config elements here

makramkd replied:
That's a good idea, why don't we make an interface like:

```golang
type ChangeSetConfig interface {
    Validate() error
}
``` 

and then the ChangeSet generic can be:

```golang
type ChangeSet[C ChangeSetConfig] func(e Environment, config C) (ChangesetOutput, error)
```

wdyt?

AnieeG replied:
nice ! yeah we can do this to enforce validation always

makramkd replied:
Created https://smartcontract-it.atlassian.net/browse/CCIP-4393

AnieeG commented on deployment/ccip/changeset/accept_ownership_test.go:
```diff
@@ -0,0 +1,136 @@
+)
+
+func Test_NewAcceptOwnershipChangeset(t *testing.T) {
+	e := NewMemoryEnvironmentWithJobs(t, logger.TestLogger(t), 2, 4)
```
I think we can just use a generic test for this if you are moving to common changeset

makramkd replied:
I kept the CCIP test but moved the changeset to common

AnieeG commented on deployment/ccip/changeset/accept_ownership.go:
```diff
@@ -0,0 +1,149 @@
+
+type AcceptOwnershipConfig struct {
+	State             CCIPOnChainState
+	ChainSelectors    []uint64
```
how about you take a struct of contract and chainselector combination here and move the changeset to commonchangeset?

AnieeG commented on deployment/ccip/changeset/transfer_ownership.go:
```diff
@@ -0,0 +1,95 @@
+
+type TransferOwnershipConfig struct {
+	State             CCIPOnChainState
+	ChainSelectors    []uint64
```
same comment as accept ownership

makramkd committed:
move changeset to common

Refactor the proposal helpers a bit

makramkd committed:
move transfer ownership cs to common

makramkd committed:
Merge branch 'develop' into mk/accept-ownership-proposal

makramkd committed:
fix

makramkd committed:
Merge branch 'develop' into mk/accept-ownership-proposal

makramkd committed:
Merge branch 'develop' into mk/accept-ownership-proposal

makramkd committed:
bump boost significantly

the AddChainInbound test is consistently failing with a "message too
costly" error in exec; increasing the relative boost per wait hour
causes exec to significantly boost the paid fee so we can execute the
message.

graham-chainlink commented on deployment/common/proposalutils/propose.go:
```diff
@@ -0,0 +1,77 @@
+	return metaDataPerChain, nil
+}
+
+// Given batches of operations, we build the metadata and timelock addresses of those opartions
```
```suggestion
// Given batches of operations, we build the metadata and timelock addresses of those operations
```
